### PR TITLE
feat: workload labels

### PR DESCRIPTION
**Pull request by bernot-dev**
Reimplement #1065 with new requirements, as revised in go/gmp:expand-target-labels.

### CODE REVIEW

**Review by dashpole — COMMENTED** [review]
looks reasonable to me, and i'll let bartek approve.  I assume there no good way to test this against actual k8s objects?

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 1]
```diff
@@ -382,6 +386,76 @@ func relabelingsForMetadata(keys map[string]struct{}) (res []*relabel.Config) {
 			TargetLabel:  "node",
 		})
 	}
+	if _, ok := keys["top_level_controller_name"]; ok {
+		res = append(res,
+			// Default to pod name
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:391]: nit: Can we use full sentence and trailing `.` for comments in this file, for clarity? thanks!

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 2]
```diff
@@ -32,17 +32,21 @@ import (
...
 	allowedClusterPodMonitoringLabel = map[string]bool{
-		"container": true,
-		"namespace": true,
-		"node":      true,
-		"pod":       true,
+		"container":                 true,
+		"namespace":                 true,
+		"node":                      true,
+		"pod":                       true,
+		"top_level_controller_name": true,
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:39]: Do you mind switching to constants in this file? Bit of work, but avoid typos 🙈

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 3]
```diff
@@ -382,6 +386,76 @@ func relabelingsForMetadata(keys map[string]struct{}) (res []*relabel.Config) {
 			TargetLabel:  "node",
 		})
 	}
+	if _, ok := keys["top_level_controller_name"]; ok {
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:389]: I would say this deserves a simple unit test that puts what we expect to see in those `__meta` labels and do the resulting relabelling. We can then track regressions or weird __meta label states over time. WDYT? (:

**Review by bwplotka — COMMENTED** [review]
Thanks! This is great and super important work.

1. Added some nits and unit test suggestions.
2. I checked the past discussions, and there is one more thing we noticed:
   * Is it worth to mention those new labels in [`TargetLabels.Metadata`](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/b1a752822b70eaf672ff69a961dbb441d5f7a749/pkg/operator/apis/monitoring/v1/pod_types.go#L304) for both cluster and pod monitorings?
3. Did you manage to self-deploy GMP with operator from this PR, on GKE std cluster? It would be epic to have at least basics like daemonset, deployment and statefulset workloads and PodMonitorings to see if the labels appear correctly? Would be epic to do this, given no e2e for this (the alternative is to write an e2e test, that would be probably even better).

Otherwise perfect. I would love to propose same labels to Prometheus SD one day (:

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 4]
```diff
@@ -382,6 +386,76 @@ func relabelingsForMetadata(keys map[string]struct{}) (res []*relabel.Config) {
...
+			&relabel.Config{
+				Action:       relabel.Replace,
+				SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name"},
+				TargetLabel:  "top_level_controller_name",
+			},
+			// For non-ReplicaSet and non-Job controllers
+			&relabel.Config{
+				Action:       relabel.Replace,
+				SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_controller_name"},
+				Regex:        relabel.MustNewRegexp("(.+)"),
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:401]: Do we need this (and `Replacement`) in this entry?

That seems to be [a Prometheus default anyway](https://github.com/prometheus/prometheus/blob/main/model/relabel/relabel.go#L32)?

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 5]
```diff
@@ -382,6 +386,76 @@ func relabelingsForMetadata(keys map[string]struct{}) (res []*relabel.Config) {
...
+				SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name"},
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				TargetLabel:  "top_level_controller_type",
+				Replacement:  "Pod",
+			},
+			// Capture controller type
+			&relabel.Config{
+				Action:       relabel.Replace,
+				SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_controller_kind"},
+				Regex:        relabel.MustNewRegexp("(.+)"),
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:437]: ditto

File: pkg/operator/apis/monitoring/v1/pod_config.go [hunk 6]
```diff
@@ -382,6 +386,76 @@ func relabelingsForMetadata(keys map[string]struct{}) (res []*relabel.Config) {
...
+	}
+	if _, ok := keys["top_level_controller_type"]; ok {
+		res = append(res,
+			// Default to Pod
+			&relabel.Config{
+				Action:       relabel.Replace,
+				SourceLabels: prommodel.LabelNames{"__meta_kubernetes_pod_name"},
+				Regex:        relabel.MustNewRegexp("(.*)"),
+				TargetLabel:  "top_level_controller_type",
+				Replacement:  "Pod",
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/pod_config.go:431]: Curious, why `pod` and not `unknown` (clear distinction that we can't / have bug in the workload discovery?) . Not a bad choice, just curious of pros & cons, can't see this explained in the design.

Same for controller_name 🤔

**Review by bwplotka — APPROVED** [review]
Wow, addressed so quickly! Amazing, LGTM, thanks 💪🏽 

Unit test would be nice for relabel again, but not strictly now in this PR!

**Comment by bwplotka** [comment]
Nice, thanks!
 
image

**Pull request merged**